code/go/0chain.net/blobbercore/challenge/challenge.go

```diff
@@ -26,186 +24,162 @@ type BCChallengeResponse struct {
-			logging.Logger.Info("[challenge]open: No challenge entity from the challenge map")
-			continue
+	for {
+		var challenges BCChallengeResponse
+		var challengeIDs []string
+		challenges.Challenges = make([]*ChallengeEntity, 0)
+		apiStart := time.Now()
+		retBytes, err := transaction.MakeSCRestAPICall(transaction.STORAGE_CONTRACT_ADDRESS, "/openchallenges", params, chain.GetServerChain())
+		if err != nil {
+			logging.Logger.Error("[challenge]open: ", zap.Error(err))
+			break
 		}
 
-		saveNewChallenge(challengeObj, ctx)
-	}
+		downloadElapsed += time.Since(apiStart)
 
-	logging.Logger.Info("[challenge]elapsed:pull",
-		zap.Int("count", len(blobberChallenges.Challenges)),
-		zap.String("download", downloadElapsed.String()),
-		zap.String("json", (jsonElapsed-downloadElapsed).String()),
-		zap.String("db", (time.Since(startTime)-downloadElapsed-jsonElapsed).String()))
-
-}
-
-func saveNewChallenge(c *ChallengeEntity, ctx context.Context) {
-	defer func() {
-		if r := recover(); r != nil {
-			logging.Logger.Error("[recover]add_challenge", zap.Any("err", r))
+		jsonStart := time.Now()
+		bytesReader := bytes.NewBuffer(retBytes)
+		d := json.NewDecoder(bytesReader)
+		d.UseNumber()
+		if err := d.Decode(&challenges); err != nil {
+			logging.Logger.Error("[challenge]json: ", zap.String("resp", string(retBytes)), zap.Error(err))
+			break
 		}
-	}()
-
-	startTime := time.Now()
-
-	if _, err := cMap.Get(c.ChallengeID); err == nil {
-		return
-	}
+		for _, c := range challenges.Challenges {
+			challengeIDs = append(challengeIDs, c.ChallengeID)
+			if c.CreatedAt > common.Timestamp(lastChallengeTimestamp) {
+				lastChallengeTimestamp = int(c.CreatedAt)
+			}
+		}
+		logging.Logger.Info("challenges_from_chain",
+			zap.Int("challenges", len(challenges.Challenges)),
+			zap.Strings("challenge_ids", challengeIDs))
 
-	db := datastore.GetStore().GetDB()
-	if status := getStatus(db, c.ChallengeID); status != nil {
-		cMap.Add(c.ChallengeID, *status) //nolint
-		return
+		jsonElapsed += time.Since(jsonStart)
+		if len(challenges.Challenges) == 0 {
+			break
+		}
+		allOpenChallenges = append(allOpenChallenges, challenges.Challenges...)
+		offset += incrOffset
+		params["offset"] = strconv.Itoa(offset)
 	}
 
-	c.Status = Accepted
-	createdTime := common.ToTime(c.CreatedAt)
-
-	logging.Logger.Info("[challenge]add: ",
-		zap.String("challenge_id", c.ChallengeID),
-		zap.Time("created", createdTime))
-
-	if err := db.Transaction(func(tx *gorm.DB) error {
-		return c.SaveWith(tx)
-	}); err != nil {
-		logging.Logger.Error("[challenge]add: ",
-			zap.String("challenge_id", c.ChallengeID),
-			zap.Time("created", createdTime),
-			zap.Error(err))
+	dbTimeStart := time.Now()
+	logging.Logger.Info("Starting saving challenges",
+		zap.Int("challenges", len(allOpenChallenges)))
 
+	if len(allOpenChallenges) == 0 {
 		return
 	}
 
-	cMap.Add(c.ChallengeID, Accepted) //nolint
+	saved := saveNewChallenges(ctx, allOpenChallenges...)
```

Comment:
Why expand slice?
We could simply send slice instead.

Reply:
any issues with above implementation?
it allows to pass a single id as well without converting to slice..